### ISSUE TITLE
Possible problems involving `clippy::shadow_reuse` and auto-generated code

### ISSUE THREAD

**Issue by c410-f3r** [rust-lang/rust-clippy#9757]
### Description

According to https://github.com/dtolnay/quote/issues/232#issuecomment-1296451876, `clippy::shadow_reuse` shouldn't be triggered but I am personally not aware of this intended limitation. Any thoughts?

### Version

```text
Latest 1.65 beta
```

**Comment by Alexendoo** [comment]
A fix for this case would be to check `pat.span.from_expansion()` in addition to the current `ident.span.from_expansion()` check

It wouldn't be intended for this to lint, ident + macro interactions are just non obvious

**Comment by c410-f3r** [comment]
Closing due to the lack of activity.

**Comment by lochetti** [comment]
I am trying to take a look at this issue. Although, this would be my first contribution here.

Not linting the code if `pat.span.from_expansion()` does solve the specific issue mentioned for the `quote` crate, but I have one doubt: Which is the best way to increment the tests in order to add this use case? I am really not sure how to do that.

Thanks for the patience in helping a new contributor wannabe!

**Comment by c410-f3r** [comment]
To solve the issue, I **think** `shadow_reuse` should also check `clippy_utils::is_from_proc_macro` along side the already included `ident.span.from_expansion()`. See https://github.com/rust-lang/rust-clippy/pull/10203#issuecomment-1509929940.

In regards to test addition or how to test this issue specifically, I really have no clue because https://github.com/rust-lang/rust-clippy/pull/10203/files#diff-98b6b061d5b311aae09fd482c612aab5544d55b9c0da67438d79ef0fe181d526R129-R159 didn't help.

Perhaps one of the Clippy members can help you? @Alexendoo @Jarcho @xFrednet

**Comment by Alexendoo** [comment]
You can add a test along the lines of

```rust
macro_rules! reuse {
    ($v:ident) => {
        let $v = $v + 1;
    }
}

let x = 1;
reuse!(x);
```
